### ISSUE TITLE
[libc++] std::deque loops over the deque for ASAN even when annotations are disabled

### ISSUE THREAD

**Issue by ldionne** [llvm/llvm-project#73043]
This seems unintended: https://github.com/llvm/llvm-project/blob/main/libcxx/include/deque#L987

**Comment by AdvenamTacet** [comment]
We rely here on compiler optimizations. The loop comprises no-op operations and is thus expected to be eliminated during the optimization process. That this optimization can happez can be confirmed by examining the assembly code. When compiled [with the -O1 flag, LLVM16 and LLVM17 produce identical output](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1DIApACYAQuYukl9ZATwDKjdAGFUtAK4sGEgKykrgAyeAyYAHI%2BAEaYxCCBAA6oCoRODB7evgGkyamOAqHhUSyx8YF2mA7pQgRMxASZPn5cFZj2BQy19QRFkTFxCbZ1DU3ZrcM9fSVlCQCUtqhexMjsHOYAzGHI3lgA1CYbblgAjl6Yh9gmGgCCm9u7mAdHyAoE%2BIKX13e3AG6oeHQe34qAgb3QIBAp3OhzcXjCBAAbJIAPoES4HMyI9BzA4Adistz2xL26AAdDtMPUIHNDoS7niACIcBa0Tj%2BXh%2BDhaUioThuazWPYKJYrJ6bHikAiaFmkCBINAsRJ0OLkSiK5X0eI7QzALh4jRcPh0AhxBSUaIy0jRML1ACenElNuYxDtAHlotoqtLuLxFWxBG6GLQHdzeFhol5gG4xLRzb7SFgWLrxGHE3hiN68L9MPGeZhVFUvKbHbwEe0rbQ8NFiPaPFhS1LiHgWKWFlQDMAFAA1PCYADubsSjEb/EEIjE7CkMkEihU6jTuiNBiMKEFln01fNkAWqESnXjAFpwVXoodGafTJYhcf3uer1ZLGYNHtD26uK/gBteKgc8Rm1g240rY7RZs4ECuGMLRBAw6DTAM8RGnkaQCFBejIZ08GlIMRqVNUAjdKMnjNHoeGdIRvRhP02GIZMRFZNBbxTFRMw4QsIrLKsEisuynJWnyHB7KoAAciKHsiew6kYez6mSGhkh%2BEC4IQJCYhsXBzLwPpaHMCwANYJBo%2BicJIvCtlwGhGVyPICbw5pGtpspwLACqoEqKpkBQEAah5KArnqiKWcatCmsQ5oQJaabOvajbRa6Hpeg4jb%2BowBBBiGVoRlGMa0HGjZJimaw8vgmbVDmea8AWRYlgm5Zsmmp61q69ZrJKBDNq2vrtp2PZ9oOw5cpKY7CKI4jTsNc5qFauhmPouprteG6nkBu77ukR4ntW56XuuFivuC967c%2Br7vp%2B368n%2BAG5vA7GgfhfgQbBaGSBsMFwSxCESLNGHpM9r0/YUH00V9IEdDUIyNMR2QbHioNgV0ENYbMXC4RDaEw3RlHFJ9KPsaKXEacZHAcqQ1k/pwQmieJkiSf5MmInJcl7Ep%2BBEMQakaVpMq6aQCBUlg8TAQZ/hGfVpmk/xnB2SARmOTz9VmHxaa2VK3MLH%2BqTOJIQA%3D%3D%3D).

This code also iterates over modified chunks only (not even elements), so it should never impact complexity. When one object is added/removed and annotations are not optimized-out, the loop is executed only once.

It should also affect only programs without optimizations (-O0) in any way.
Annotations should not end up in optimized binaries.

When compiled without optimizations (-O0), based on one oversimplified benchmark I was able to run, the performance impact of this code is negligible. (I compared the performance of LLVM16 and LLVM17.)

However, that amount of additional code with -O0 is not desirable, I can prepare a patch on Monday minimizing it (but not fully removing). But should I do the same for the std::vector container? While vector adds much less calculations, the logic is the same: `annotate_*` is called and some calculations are made. That would double the amount of annotating functions.

The rationale behind maintaining a single set of functions and relying on compiler optimizations lies in the avoidance of redundant code maintenance and minimization of conditional code guarded by #ifdef directives. These optimizations are relatively straightforward, and I anticipate encountering no instances where annotations persist during optimization.

In summary, I propose retaining the current implementation unless there are instances where these functions fail to undergo optimization. However, I acknowledge that `__annotate_from_to` is more extensive than the other annotating functions, potentially warranting an exception. While I currently favor maintaining the existing code, I am open to creating separate versions of `__annotate_from_to` or all functions if deemed necessary. I will conduct further research and provide an update early next week.

**Comment by ldionne** [comment]
Thanks a lot for your analysis. Would it be possible to do this instead?

```
_LIBCPP_HIDE_FROM_ABI
void __annotate_from_to([[maybe_unused]] size_type __beg, 
                        [[maybe_unused]] size_type __end, 
                        [[maybe_unused]] __asan_annotation_type __annotation_type, 
                        [[maybe_unused]] __asan_annotation_place __place) const _NOEXCEPT {
#if !defined(_LIBCPP_HAS_NO_ASAN)
    // __beg - index of the first item to annotate
    // __end - index behind the last item to annotate (so last item + 1)
    // __annotation_type - __asan_unposion or __asan_poison
    // __place - __asan_front_moved or __asan_back_moved
    // Note: All indexes in __map_
    if (__beg == __end)
        return;

    ...

#endif
}
```

The same approach could be done for `__annotate_double_ended_contiguous_container` in `deque` but also for the other functions in `vector`. This would not require having multiple versions of each function but there would indeed be some more `#ifdef`s in the code. In this case I think this is reasonable though, I don't think we should be relying on the compiler's ability to remove dead code that much in such widely used classes.

**Comment by AdvenamTacet** [comment]
I wasn't aware that `[[maybe_unused]]` exists, that resolves the problem of duplicated functions.

That should work. I can test it and create the PR (Monday at the latest). I agree it's reasonable.

I will do it in deque and vector (and string PR) for functions with two versions, as well as `__annotate_from_to`, because it has a lot of code inside.

If you think that we should do that for every `__annotate...` function, I can create a separate PR, but I think we should ask more people for feedback. Throughout the development of these patches, I received feedback from 2+ persons, asking to minimize the code guarded by macros (it wasn't my idea). And all other functions have minimal code inside, often even only a single call, so I'm unsure about benefits.

**Comment by AdvenamTacet** [comment]
I sent a commit with that kind of change to the string annotations PR and it looks good.

https://github.com/llvm/llvm-project/pull/72677/commits/6da17753d4377af4c5d6b35ecedbe02f9f89bbba

**Comment by AdvenamTacet** [comment]
Hey, I created #74023 based on suggestion from discord (`[[__maybe_unused__]]` instead of `[[maybe_unused]]`, thx @philnik777), it seems to work with C++03, the failed test seems to be unrelated. Seems ready for the review, I will rebase the commit soon and hopefully it fixes the test case fail.
